### PR TITLE
Genutil xmod | collect-gentxs

### PR DESCRIPTION
**Pull request by NYBACHOK**

### CODE REVIEW

File: gaia-rs/src/store_keys.rs [hunk 1]
```diff
@@ -16,6 +16,8 @@ pub enum GaiaStoreKey {
     IBC,
     #[skey(to_string = "capability")]
     Capability,
+    #[skey(to_string = "genutil")]
+    Genutil,
```
Comment (NYBACHOK) [inline gaia-rs/src/store_keys.rs:20]: Is `app_state` contains all StoreKeys or it smth different?

File: x/genutil/src/balances_iter.rs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+impl GenesisBalanceIter {
+    pub fn new<SK: StoreKey>(sk: &SK, genesis_path: impl AsRef<Path>) -> anyhow::Result<Self> {
+        let mut value: serde_json::Value = serde_json::from_slice(&std::fs::read(genesis_path)?)?;
+
+        let value = value
+            .get_mut("app_state")
+            .ok_or(anyhow::anyhow!("missing `app_state`"))?
+            .get_mut(sk.name())
+            .ok_or(anyhow::anyhow!("module is not found"))?
+            .get_mut("balances")
```
Comment (NYBACHOK) [inline x/genutil/src/balances_iter.rs:26]: I'm slightly worried about this part `.get_mut("balances")`

**Review by dmvict — APPROVED** [review]

File: gaia-rs/src/store_keys.rs (hunk 1 above)
Comment (joneskm) [reply gaia-rs/src/store_keys.rs:20]: From what I can see in Gaia genutils doesn't have an entry in the store

File: x/genutil/src/balances_iter.rs (hunk 2 above)
Comment (joneskm) [reply x/genutil/src/balances_iter.rs:26]: Why does genutils care about bank balances?

File: x/genutil/src/client/cli/collect.rs [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        CollectGentxCliAux {
+            gentx_dir,
+            home,
+            moniker,
+            _marker,
+        }: CollectGentxCliAux<AI>,
+    ) -> Result<Self, Self::Error> {
+        if !gentx_dir.exists() {
+            Err(anyhow::anyhow!(
+                "Failed to find folder: {}",
```
Comment (joneskm) [inline x/genutil/src/client/cli/collect.rs:34]: Directory instead of folder?

File: x/staking/src/message.rs [hunk 4]
```diff
@@ -10,7 +10,7 @@ use serde::Serialize;
 #[serde(tag = "@type")]
 #[allow(clippy::large_enum_variant)]
 pub enum Message {
-    #[serde(rename = "/cosmos.staking.v1beta1.CreateValidator")]
+    #[serde(rename = "/cosmos.staking.v1beta1/MsgCreateValidator")]
```
Comment (joneskm) [inline x/staking/src/message.rs:13]: Why the extra `/`

File: x/staking/src/types/tx.rs [hunk 5]
```diff
@@ -347,6 +349,46 @@ pub struct CreateValidator {
     pub value: UnsignedCoin,
 }
 
+impl CreateValidator {
+    pub const TYPE_URL: &'static str = "/cosmos.staking.v1beta1/MsgCreateValidator";
```
Comment (joneskm) [inline x/staking/src/types/tx.rs:353]: Why the second `/`?

File: x/staking/src/types/tx.rs [hunk 6]
```diff
@@ -347,6 +349,46 @@ pub struct CreateValidator {
...
+        Self::TYPE_URL
+    }
+}
+
+impl TryFrom<Any> for CreateValidator {
+    type Error = CoreError;
+
+    fn try_from(value: Any) -> Result<Self, Self::Error> {
+        match value.type_url.as_str() {
+            "/cosmos.staking.v1beta1.CreateValidator" => {
```
Comment (joneskm) [inline x/staking/src/types/tx.rs:371]: `MsgCreateValidator`

File: gaia-rs/src/store_keys.rs (hunk 1 above)
Comment (NYBACHOK) [reply gaia-rs/src/store_keys.rs:20]: But it has to be in genesis. Okay, thanks.

File: x/genutil/src/balances_iter.rs (hunk 2 above)
Comment (NYBACHOK) [reply x/genutil/src/balances_iter.rs:26]: 'cause it needs to read balances of delegator. [Usage](https://github.com/NYBACHOK/gears/blob/6c8074953be648fa1d85a869cb761f9fc8c11f10/x/genutil/src/collect_txs.rs#L131-L141)

File: x/genutil/src/client/cli/collect.rs (hunk 3 above)
Comment (NYBACHOK) [reply x/genutil/src/client/cli/collect.rs:34]: Oh. I need to change it.

File: x/staking/src/message.rs (hunk 4 above)
Comment (NYBACHOK) [reply x/staking/src/message.rs:13]: My mistake. I thought that it same as for query

File: x/staking/src/types/tx.rs (hunk 5 above)
Comment (NYBACHOK) [reply x/staking/src/types/tx.rs:353]: Comment above

File: x/staking/src/types/tx.rs (hunk 6 above)
Comment (NYBACHOK) [reply x/staking/src/types/tx.rs:371]: Okay

File: gaia-rs/src/store_keys.rs (hunk 1 above)
Comment (NYBACHOK) [reply gaia-rs/src/store_keys.rs]: Removed

File: x/genutil/src/balances_iter.rs (hunk 2 above)
Comment (joneskm) [reply x/genutil/src/balances_iter.rs:26]: OK

**Pull request merged**